src/EnergyPlus/WaterThermalTanks.cc

```diff
@@ -8410,6 +8429,487 @@ namespace WaterThermalTanks {
+            if (Tank.HeaterOn1 || Tank.HeaterOn2) Runtime += dt;
+            Eheater1 += Qheater1 * dt;
+            Eheater2 += Qheater2 * dt;
+            Eunmet += max(Tank.UseEffectiveness * Tank.UseMassFlowRate * (UseOutletExpectedTemp - Tavg[Tank.UseOutletStratNode - 1]), 0.0) * dt;
+
+        } // end while TimeRemaining > 0.0
+
+        for (auto &e : Tank.Node) {
+            e.TempAvg = e.TempSum / SecInTimeStep;
+            e.TempSum = 0.0;
+        }
+
+        Tank.TankTemp = sum(Tank.Node, &StratifiedNodeData::Temp) / Tank.Nodes;
+        Tank.TankTempAvg = sum(Tank.Node, &StratifiedNodeData::TempAvg) / Tank.Nodes;
+
+        if (Tank.UseOutletStratNode > 0) {
+            Tank.UseOutletTemp = Tank.Node(Tank.UseOutletStratNode).TempAvg;
+            // Revised use outlet temperature to ensure energy balance. Assumes a constant CP. CR8341/CR8570
+            if (Tank.UseMassFlowRate > 0.0) {
+                Tank.UseOutletTemp = Tank.UseInletTemp * (1.0 - Tank.UseEffectiveness) + Tank.UseOutletTemp * Tank.UseEffectiveness;
+            }
+        }
+
+        if (HPWHCondenserConfig == TypeOf_HeatPumpWtrHeaterWrapped) {
+            // If we have a wrapped condenser HPWH, set the source outlet to the weighted average of the node
+            // temperatures the condenser sees
+            Real64 WeightedAverageSourceOutletTemp(0.0);
+            for (int i = 1; i <= Tank.Nodes; ++i) {
+                WeightedAverageSourceOutletTemp += Tank.Node(i).TempAvg * Tank.Node(i).HPWHWrappedCondenserHeatingFrac;
+            }
+            Tank.SourceOutletTemp = WeightedAverageSourceOutletTemp;
+        } else if (Tank.SourceOutletStratNode > 0) {
+            // otherwise set it to the temperature of the source outlet node
+            Tank.SourceOutletTemp = Tank.Node(Tank.SourceOutletStratNode).TempAvg;
+        }
+        if (HPWHCondenserConfig == TypeOf_HeatPumpWtrHeaterPumped) {
+            // For pumped condensers, set the source inlet and outlets to match the delta T
+            // across the water side of the DX coil.
+            HeatPumpWaterHeaterData const &HeatPump = HPWaterHeater(Tank.HeatPumpNum);
+            DataLoopNode::NodeData const &HPWHCondWaterInletNode = DataLoopNode::Node(HeatPump.CondWaterInletNode);
+            DataLoopNode::NodeData const &HPWHCondWaterOutletNode = DataLoopNode::Node(HeatPump.CondWaterOutletNode);
+            Real64 const HPWHCondenserDeltaT = HPWHCondWaterOutletNode.Temp - HPWHCondWaterInletNode.Temp;
+            Tank.SourceInletTemp = Tank.SourceOutletTemp + HPWHCondenserDeltaT;
+        }
+
+        // Revised source outlet temperature to ensure energy balance. Assumes a constant CP. CR8341/CR8570
+        if (Tank.SourceOutletStratNode > 0) {
+            if (Tank.SourceMassFlowRate > 0.0) {
+                Tank.SourceOutletTemp = Tank.SourceInletTemp * (1.0 - Tank.SourceEffectiveness) + Tank.SourceOutletTemp * Tank.SourceEffectiveness;
+            }
+        }
```

Comment:
Can this be moved up into the else if at line 8854?

Reply:
i.e., can line 8870-8872 be moved up to 8857.

Reply:
No, because the source inlet temperature could be changed on line 8865.